services/repository/contributors_graph.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package repository
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	repo_model "code.gitea.io/gitea/models/repo"
+	user_model "code.gitea.io/gitea/models/user"
+	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/log"
+	api "code.gitea.io/gitea/modules/structs"
+	util "code.gitea.io/gitea/modules/util"
+)
+
+// CreateWeeks converts list of sundays to list of *api.WeekData
+func CreateWeeks(sundays []int64) []*api.WeekData {
+	var weeks []*api.WeekData
+	for _, week := range sundays {
+		weeks = append(weeks, &api.WeekData{
+			Week:      week,
+			Additions: 0,
+			Deletions: 0,
+			Commits:   0,
+		},
+		)
+	}
+	return weeks
+}
+
+// GetContributorStats returns contributors stats for git commits for given revision or default branch
+func GetContributorStats(ctx context.Context, repo *repo_model.Repository, revision string) (map[string]*api.ContributorData, error) {
+	gitRepo, closer, err := git.RepositoryFromContextOrOpen(ctx, repo.RepoPath())
+	if err != nil {
+		return nil, fmt.Errorf("OpenRepository: %w", err)
+	}
+	defer closer.Close()
+
+	if len(revision) == 0 {
+		revision = repo.DefaultBranch
+	}
+	extendedCommitStats, err := gitRepo.ExtendedCommitStats(revision)
+	if err != nil {
+		return nil, fmt.Errorf("ExtendedCommitStats: %w", err)
+	}
+
+	layout := time.DateOnly
+	initialCommitDate := extendedCommitStats[0].Author.Date
+
+	startingSunday, _ := util.FindLastSundayBeforeDate(initialCommitDate)
+	endingSunday, _ := util.FindFirstSundayAfterDate(time.Now().Format(layout))
+
+	sundays, _ := util.ListSundaysBetween(startingSunday, endingSunday)
+
+	unknownUserAvatarLink := user_model.NewGhostUser().AvatarLink(ctx)
+	contributorsCommitStats := make(map[string]*api.ContributorData)
+	contributorsCommitStats["total"] = &api.ContributorData{
+		Name:       "Total",
+		AvatarLink: unknownUserAvatarLink,
+		Weeks:      CreateWeeks(sundays),
+	}
+	total := contributorsCommitStats["total"]
+
+	for _, v := range extendedCommitStats {
+		userEmail := v.Author.Email
+		if len(userEmail) == 0 {
+			continue
+		}
+		u, _ := user_model.GetUserByEmail(ctx, userEmail)
+		if u != nil {
+			// update userEmail with user's primary email address so
+			// that different mail addresses will linked to same account
+			userEmail = u.GetEmail()
+		}
+		if _, ok := contributorsCommitStats[userEmail]; !ok {
+			if u == nil {
+				contributorsCommitStats[userEmail] = &api.ContributorData{
+					Name:       v.Author.Name,
+					AvatarLink: unknownUserAvatarLink,
+					Weeks:      CreateWeeks(sundays),
+				}
```

Comment:
```suggestion
				avatarLink := avatars.GenerateEmailAvatarFastLink(ctx, userEmail, 0)
				if avatarLink == "" {
					avatarLink = unknownUserAvatarLink
				}

				contributorsCommitStats[userEmail] = &api.ContributorData{
					Name:       v.Author.Name,
					AvatarLink: avatarLink,
					Weeks:      CreateWeeks(sundays),
				}
```
You need to add some imports

Before:
![image](https://github.com/go-gitea/gitea/assets/18380374/7086a696-62a5-4882-88a3-dde55b15e4e6)

After:
![image](https://github.com/go-gitea/gitea/assets/18380374/c5ee9e2e-4dc2-4c9c-b4b6-b5262ce37ab2)

The avatars will be as same as what are listed in the commits list page.
![image](https://github.com/go-gitea/gitea/assets/18380374/d5af6a63-25aa-4e21-b902-10e13eef4cac)

@silverwind 
How do you think about this change? I'm not sure about the parameter `size== 0` for `GenerateEmailAvatarFastLink`

Reply:
Maybe we need a function for Vue in order to get the same usage in the template like:
`{{ctx.AvatarUtils.AvatarByEmail .Commit.Author.Email .Commit.Author.Email 28 "gt-mr-3"}}`

Reply:
I've created avatars by following your advice. It's looking much better, thanks!